src/core/vectortile/qgsvectortilelayerrenderer.cpp

```diff
@@ -108,7 +108,34 @@ bool QgsVectorTileLayerRenderer::render()
   tTotal.start();
 
   const double tileRenderScale = mTileMatrixSet.scaleForRenderContext( ctx );
-  QgsDebugMsgLevel( QStringLiteral( "Vector tiles rendering extent: " ) + ctx.extent().toString( -1 ), 2 );
+
+  QgsRectangle extent = ctx.extent();
+  if ( extent.isMaximal() )
+  {
+    // invalid extent. We don't want to fetch ALL tiles at the zoom level!
+    // This has occurred because the map renderer is being pessimistic and thinks a worst-case
+    // scenario is acceptable when it failed to transform the map extent to the layer's crs.
+    // But while that's permissible eg for a local raster layer, it's entirely inappropriate
+    // for vector tiles!
+    // So let's try and determine a MINIMAL extent for the layer, avoiding the pessimism used
+    // in the generic map renderer code
```

Comment:
I am not sure I fully understand how this works... the layer extent given by map renderer has at least one value going to infinity, so we take the map extent and try to transform it to layer extent with some more relaxed options? what does the coordinate transform return instead?

Reply:
We are hitting this branch in the generic code - https://github.com/qgis/QGIS/blob/master/src/core/maprenderer/qgsmaprendererjob.cpp#L415

I don't know if it's safe to remove that generic logic or not. It's quite possible that it's redundant and note required anymore due to other changes in bounding box transformation logic. I'd rather be cautious and just workaround it in the individual layer renderer subclasses as required (like we do here)

Reply:
I see... thanks for the explanation!